junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertTimeout.java

```diff
@@ -180,4 +182,30 @@ static void assertTimeoutPreemptively(Duration timeout, Executable executable, S
 		}
 	}
 
+	/**
+	 * The thread factory used for preemptive timeout.
+	 *
+	 * This factory is the same with {@link Executors#defaultThreadFactory()} but provides more meaningful thread names,
+	 * helpful for debugging purposes.
+	 *
+	 */
+	private static class TimeoutThreadFactory implements ThreadFactory {
+		private static final AtomicInteger threadNumber = new AtomicInteger(1);
+		private final ThreadGroup group;
+
+		TimeoutThreadFactory() {
+			SecurityManager s = System.getSecurityManager();
+			group = (s != null) ? s.getThreadGroup() : Thread.currentThread().getThreadGroup();
+		}
+
+		public Thread newThread(Runnable r) {
+			Thread t = new Thread(group, r, "junit-timeout-thread-" + threadNumber.getAndIncrement(), 0);
+			if (t.isDaemon())
+				t.setDaemon(false);
+			if (t.getPriority() != Thread.NORM_PRIORITY)
+				t.setPriority(Thread.NORM_PRIORITY);
```

Comment:
I don't think we need these two `if`s as we shut down the executor afterwards.

Reply:
I don't think it matters if we shutdown the executor or not. Even after shutdown the thread is not guaranteed to terminate so being a daemon or not matters. 

Moreover, in a setting where there many threads competing for CPU if these ones have low/high priority it might have an impact.

Reply:
> Even after shutdown the thread is not guaranteed to terminate so being a daemon or not matters.

D you mean we should make it a daemon thread to avoid it from preventing the VM to shut down?

Reply:
Basically, I am saying that we should decide if we want daemon or non daemon threads. If we don't specify it then we rely on the status of the parent thread. In this case, I could prefer non-daemon threads.